thefuck/rules/cat_dir.py

```diff
@@ -0,0 +1,12 @@
+import re
+
+
+def match(command):
+    return (
+        command.script.startswith('cat') and
+        re.match(r'cat: .+: Is a directory', command.output)
```

Comment:
`re.match` could be replaced with `str.startswith`, `str.rstrip` and `str.endswith` used in conjunction. `str.rstrip` is needed here because `command.output` ends with `\n` (see previous line comment)

Reply:
What are the relative merits of `command.output.rstrip().endswith(': Is a directory')` vs. `command.output.endswith(': Is a directory\n')`?

Reply:
Well, as you probably guessed, my initial thought was that there may be a version – or a user defined alias or function – of `cat` that doesn't add that trailing newline. Then I went on and did some benchmarks and it turns out `rstrip` makes it around 20% slower – not that much. Maybe it's worth having.

Or maybe we're better off with just `... endswith(': Is a directory\n')` and those who use a `cat` that doesn't follow “default” implementation are ”in a state of sin and deserve whatever happens to them.” We can still have ` ... or endswith(': Is a directory')` if we ever need to, after all.

I wonder what your thoughts on this are.

Reply:
Here's some of my own benchmarks
```python
# bench.py
import timeit

# the base case
rstrip_timer = timeit.Timer(r"'cat: foo: Is a directory\n'.rstrip().endswith(': Is a directory')")
# what if there's no newline?
rstrip_timer2 = timeit.Timer(r"'cat: foo: Is a directory'.rstrip().endswith(': Is a directory')")
# what if it's a mess?
rstrip_timer3 = timeit.Timer(r"'cat: foo: Is a directory \n\n \n'.rstrip().endswith(': Is a directory')")
# base case assuming one newline
no_rstrip_timer = timeit.Timer(r"'cat: foo: Is a directory\n'.endswith(': Is a directory\n')")

timers = [rstrip_timer, rstrip_timer2, rstrip_timer3, no_rstrip_timer]

try:
    number = max(timer.autorange()[0] for timer in timers)
except AttributeError:  # no autorange in Python 2
    number = 2000000
print('number = {:,}'.format(number))

results = [min(timer.repeat(repeat=5, number=number)) for timer in timers]
print('results = {}'.format(results))
results_percentages = ['{:.2%}'.format((result-results[0])/results[0]) for result in results]
print('percentages = {}'.format(results_percentages))  # bigger negative is better
```

```fish
fish $ for version in 3.7 3.6 2.7
           echo "Python $version"
           eval python$version bench.py
       end
Python 3.7
number = 2,000,000
results = [0.46968697199999987, 0.41365929999999995, 0.48141546300000027, 0.3026672379999997]
percentages = ['0.00%', '-11.93%', '2.50%', '-35.56%']
Python 3.6
number = 10,000,000
results = [2.472023139009252, 2.1997297610214446, 2.5396445730002597, 1.6017803400172852]
percentages = ['0.00%', '-11.02%', '2.74%', '-35.20%']
Python 2.7
number = 2,000,000
results = [0.48915696144104004, 0.4453470706939697, 0.563338041305542, 0.329495906829834]
percentages = ['0.00%', '-8.96%', '15.17%', '-32.64%']
```

It's interesting that Python 3.6 picks such a bigger number in autorange, but the relative timing differences are approximately the same in all Python versions. Also interesting is that the case with all kinds of screwed up whitespace is *super* slow in Python 2.7.

I think that 35% lower times are a win, since the match code is (at least in my guess) the most performance-sensitive part of thefuck.

I think using a custom alias/wrapper is both "sinful" enough and unlikely enough to not be worth supporting in the face of this performance difference. I would thus vote for the non-`rstrip` version unless anyone can find a default implementation of `cat` that outputs different whitespace or someone comes back in the future complaining that the rule doesn't work for their `cat`.

Reply:
WOW! Super! This is how far we go for open source! 🎉

Let me share my benchmarks with you:

Name (time in ns)    |        Min          |        Max          |       Mean          | StdDev          |     Median          |    IQR          | Outliers  OPS (Kops/s)          | Rounds | Iterations
---------------------|---------------------|---------------------|---------------------|-----------------|---------------------|-----------------|---------------------------------|--------|-----------
test_with_endswith   |   422.3146 (1.00)   |   441.1720 (1.02)   |   428.6322 (1.01)   | 7.4089 (4.96)   |   426.0483 (1.00)   | 7.9161 (4.70)   |      1;0    2,333.0027 (0.99)   |      5 |   10000000
test_with_rstrip     |   528.8423 (1.25)   |   532.5924 (1.24)   |   531.3621 (1.25)   | 1.4928 (1.0)    |   531.7795 (1.25)   | 1.6846 (1.0)    |      1;0    1,881.9557 (0.80)   |      5 |    1895592
test_with_match      | 1,164.0312 (2.76)   | 1,171.8660 (2.72)   | 1,167.8383 (2.74)   | 3.7479 (2.51)   | 1,166.3237 (2.75)   | 6.9720 (4.14)   |      3;0      856.2829 (0.36)   |      5 |    1000000

Generated from:

```python
import pytest
import re


class TestEndswithRstripMatch(object):

    text = 'cat: /foo/bar/: Is a directory\n'

    def with_match(self):
        return re.match(r'cat: .+: Is a directory', self.text)

    @pytest.mark.benchmark(group='endswith vs rstrip vs match')
    def test_with_match(self, benchmark):
        result = benchmark(self.with_match)
        assert result

    def with_rstrip(self):
        return self.text.startswith('cat: ') and self.text.rstrip().endswith(': Is a directory')

    @pytest.mark.benchmark(group='endswith vs rstrip vs match')
    def test_with_rstrip(self, benchmark):
        result = benchmark(self.with_rstrip)
        assert result

    def with_endswith(self):
        return self.text.startswith('cat: ') and self.text.endswith(': Is a directory\n')

    @pytest.mark.benchmark(group='endswith vs rstrip vs match')
    def test_with_endswith(self, benchmark):
        result = benchmark(self.with_endswith)
        assert result
```

You might as well need `$ pip install pytest-benchmark` and [these docs](http://pytest-benchmark.readthedocs.io/en/latest/usage.html#commandline-options).

So, sure. Let's go for the `rstrip`less version and make it more tolerant in the future, if necessary.

Reply:
Sounds good! I'm happy with my changes, so unless there's anything else you or other collaborators see needing to be done, I think I'm ready for the merge. Let me know if you would like the commits squashed.